customconfig/data_source_vault_token.go

```diff
@@ -98,38 +100,57 @@ func vaultTokenDataSourceRead(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	addr := os.Getenv("VAULT_ADDR")
+
+	var dataJSON string
+	var secret map[string]interface{}
+	err = retry(func() error {
+		dataJSON, secret, err := vaultTokenDataSourceReadCall(requestBody, addr, backend, debug)
+		if debug {
+			log.Printf("[DEBUG] Received dataJSON %s", dataJSON)
+			log.Printf("[DEBUG] Received secret %+v", secret)
+		}
+		return err
+	}, config.TimeoutMinutes)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(secret["request_id"].(string))
+	d.Set("data_json", dataJSON)
+	d.Set("lease_duration", secret["lease_duration"].(float64))
+	d.Set("lease_renewable", secret["renewable"].(bool))
+	d.Set("lease_start_time", time.Now().Format("RFC3339"))
+
+	auth := secret["auth"].(map[string]interface{})
+
+	d.Set("token", auth["client_token"].(string))
+
+	return nil
+}
+
+func vaultTokenDataSourceReadCall(requestBody []byte, addr, backend string, debug bool) (string, map[string]interface{}, error) {
+	var secret map[string]interface{}
 	resp, err := http.Post(fmt.Sprintf("%s/v1/auth/%s/login", addr, backend), "application/json", bytes.NewBuffer(requestBody))
 	if err != nil {
-		return fmt.Errorf("error talking to Vault: %s", err)
+		return "", secret, fmt.Errorf("error talking to Vault: %s", err)
 	}
 
 	defer resp.Body.Close()
 
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return fmt.Errorf("error reading from Vault: %s", err)
+		return "", secret, fmt.Errorf("error reading from Vault: %s", err)
 	}
 
 	data := string(body)
 	if debug {
 		log.Printf("[DEBUG] Got %s from Vault", string(data))
 	}
-	d.Set("data_json", string(data))
 
-	var secret map[string]interface{}
 	json.Unmarshal([]byte(data), &secret)
```

Comment:
Realize it's not part of this change, but could use `json.Unmarshal(body, &secret)` here instead of recasting the data.